[CycleAnalysis] Methods to verify cycles and their nesting.

ssahasra opened the pull request:
The original implementation provided a simple method to check whether the forest of nested cycles is well-formed. This is now augmented with other methods to check well-formedness of all cycles, either invdividually, or as the entire forest. These will be used by future transforms that modify CycleInfo.

arsenm commented on llvm/include/llvm/ADT/GenericCycleImpl.h:
```diff
@@ -119,6 +119,101 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
+        dbgs() << *BB << "\n";
+      }
+    }
+    assert(false && "Unreachable block in cycle");
```
llvm_unreachable

ssahasra replied:
Doesn't really matter here. The whole function is protected by NDEBUG, and meant to be executed only in a debug build. Also, the pattern here is "if (!condition) { print-stuff(); assert(); }" it's not really saying that a particular path is never taken, just that we need to print things before we assert the condition.

arsenm replied:
It's always a stylistic choice, and llvm_unreachable is nicer

ssahasra replied:
Fixed.

arsenm commented on llvm/include/llvm/ADT/GenericCycleImpl.h:
```diff
@@ -119,6 +119,101 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
+    dbgs() << "The following blocks are unreachable in the cycle: ";
+    for (auto *BB : Blocks) {
+      if (!VisitedBBs.count(BB)) {
+        dbgs() << *BB << "\n";
```
Probably should use printAsOperand, and use space separators? Printing the whole block is a lot

ssahasra replied:
Done.

ssahasra committed:
[CycleAnalysis] Methods to verify cycles and their nesting.

The original implementation provided a simple method to check whether the forest
of nested cycles is well-formed. This is now augmented with other methods to
check well-formedness of every cycle, either invdividually, or as the entire
forest. These will be used by future transforms that modify CycleInfo.

arsenm commented on llvm/include/llvm/ADT/GenericCycleImpl.h:
```diff
@@ -119,6 +120,104 @@ auto GenericCycle<ContextT>::getCyclePredecessor() const -> BlockT * {
+  // Setup for using a depth-first iterator to visit every block in the cycle.
+  SmallVector<BlockT *, 8> ExitBBs;
+  getExitBlocks(ExitBBs);
+  df_iterator_default_set<BlockT *> VisitSet;
+  VisitSet.insert(ExitBBs.begin(), ExitBBs.end());
```
I assume you can do this insert directly in the constructor

ssahasra replied:
That constructor exists in SmallPtrSet, but turns out that it is not exposed in df_iterator, which is a derived class. So have to go with the default constructor instead.

ssahasra committed:
addressed comment: use llvm_unreachable
